library/X509/ProvidedHook/ServicesImportSource.php

```diff
@@ -9,36 +9,69 @@ class ServicesImportSource extends x509ImportSource
-            ->groupBy(['t.ip', 't.hostname', 't.port']);
+        if ($this->getDb()->getConfig()->db === 'pgsql') {
+            $targets = (new Sql\Select())
+                ->from(['x509_target t'])
+                ->columns([
+                    'host_ip'               => 'encode(t.ip, \'hex\')',
+                    'host_name'             => 't.hostname',
+                    'host_port'             => 't.port',
+                    'cert_subject'          => 'c.subject',
+                    'cert_issuer'           => 'c.issuer',
+                    'cert_self_signed'      => 'COALESCE(ci.self_signed, c.self_signed)',
+                    'cert_trusted'          => 'c.trusted',
+                    'cert_valid_from'       => 'c.valid_from',
+                    'cert_valid_to'         => 'c.valid_to',
+                    'cert_fingerprint'      => 'encode(c.fingerprint, \'hex\')',
+                    'cert_dn'               => 'array_to_string(array_agg(CONCAT(dn.key, \'=\', dn.value)), \',\')',
+                    'cert_subject_alt_name' => (new Sql\Select())
+                        ->from('x509_certificate_subject_alt_name can')
+                        ->columns('array_to_string(array_agg(CONCAT(can.type, \':\', can.value)), \',\')')
+                        ->where(['can.certificate_id = c.id'])
+                        ->groupBy(['can.certificate_id'])
+                ])
+                ->join('x509_certificate_chain cc', 'cc.id = t.latest_certificate_chain_id')
+                ->join('x509_certificate_chain_link ccl', 'ccl.certificate_chain_id = cc.id')
+                ->join('x509_certificate c', 'c.id = ccl.certificate_id')
+                ->joinLeft('x509_certificate ci', 'ci.subject_hash = c.issuer_hash')
+                ->joinLeft('x509_dn dn', 'dn.hash = c.subject_hash')
+                ->where(['ccl.order = ?' => 0])
+                ->groupBy(['t.ip', 't.hostname', 't.port', 'c.id', 'ci.id']);
+        } else {
+            $targets = (new Sql\Select())
+                ->from('x509_target t')
+                ->columns([
+                    'host_ip'               => 't.ip',
+                    'host_name'             => 't.hostname',
+                    'host_port'             => 't.port',
+                    'cert_subject'          => 'c.subject',
+                    'cert_issuer'           => 'c.issuer',
+                    'cert_self_signed'      => 'COALESCE(ci.self_signed, c.self_signed)',
+                    'cert_trusted'          => 'c.trusted',
+                    'cert_valid_from'       => 'c.valid_from',
+                    'cert_valid_to'         => 'c.valid_to',
+                    'cert_fingerprint'      => 'HEX(c.fingerprint)',
+                    'cert_dn'               => 'GROUP_CONCAT(CONCAT(dn.key, \'=\', dn.value) SEPARATOR \',\')',
+                    'cert_subject_alt_name' => (new Sql\Select())
+                        ->from(['x509_certificate_subject_alt_name can'])
```

Comment:
The array here is redundant. That the implementation of `from()` does it isn't a reason why we should do it here. Please undo this change.